example/particle_tracers/particle_tracers.cpp

```diff
@@ -0,0 +1,515 @@
+
+  auto x1flux = rc->Get("density").flux[X1DIR].Get<4>();
+
+  // Spatially first order upwind method
+  pmb->par_for(
+      "CalculateFluxesX1", kb.s, kb.e, jb.s, jb.e, ib.s - 1, ib.e + 1,
+      KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        // X1
+        if (vmatx > 0.) {
+          x1flux(0, k, j, i) = density(k, j, i - 1) * vmatx;
+        } else {
+          x1flux(0, k, j, i) = density(k, j, i) * vmatx;
+        }
+      });
+
+  if (ndim > 1) {
+    auto x2flux = rc->Get("density").flux[X2DIR].Get<4>();
+    pmb->par_for(
+        "CalculateFluxesX2", kb.s, kb.e, jb.s, jb.e, ib.s - 1, ib.e + 1,
+        KOKKOS_LAMBDA(const int k, const int j, const int i) {
+          // X2
+          if (vmaty > 0.) {
+            x2flux(0, k, j, i) = density(k, j, i - 1) * vmaty;
+          } else {
+            x2flux(0, k, j, i) = density(k, j, i) * vmaty;
+          }
+        });
+  }
+
+  if (ndim > 2) {
+    auto x3flux = rc->Get("density").flux[X3DIR].Get<4>();
+    pmb->par_for(
+        "CalculateFluxesX3", kb.s, kb.e, jb.s, jb.e, ib.s - 1, ib.e + 1,
+        KOKKOS_LAMBDA(const int k, const int j, const int i) {
+          // X3
+          if (vmatz > 0.) {
+            x3flux(0, k, j, i) = density(k, j, i - 1) * vmatz;
+          } else {
+            x3flux(0, k, j, i) = density(k, j, i) * vmatz;
+          }
+        });
+  }
+
+  return TaskStatus::complete;
+}
+
+TaskStatus Defrag(MeshBlock *pmb) {
+  auto s = pmb->swarm_data.Get()->Get("tracers");
+
+  // Only do this if list is getting too sparse. This criterion (whether there
+  // are *any* gaps in the list) is very aggressive
+  if (s->GetNumActive() <= s->GetMaxActiveIndex()) {
+    s->Defrag();
+  }
+
+  return TaskStatus::complete;
+}
+
+// Mark all MPI requests as NULL / initialize boundary flags.
+TaskStatus InitializeCommunicationMesh(const BlockList_t &blocks) {
+  // Boundary transfers on same MPI proc are blocking
+  for (auto &block : blocks) {
+    auto swarm = block->swarm_data.Get()->Get("tracers");
+    for (int n = 0; n < block->pbval->nneighbor; n++) {
+      NeighborBlock &nb = block->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.req_send[nb.bufid] = MPI_REQUEST_NULL;
+    }
+  }
+
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("tracers");
+
+    for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+      auto &nb = pmb->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+    }
+  }
+
+  // Reset boundary statuses
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("tracers");
+    for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+      auto &nb = pmb->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+    }
+  }
+
+  return TaskStatus::complete;
+}
+
+} // namespace Particles
```

Comment:
Could these tasks be moved into the main Parthenon infrastructure? They could take a swarm name maybe as input?

Reply:
Yes I think that should be done. Right now I think the thing to do (this issue shows up in other particles examples) is merge this in to @pgrete's branch, and then make that change there everywhere all at once.

Reply:
Works for me